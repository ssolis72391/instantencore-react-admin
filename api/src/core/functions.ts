import { HasId, PartialWithRequired } from "api-shared";
import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { Mapper } from "./core";
import {
  DestroyOneById,
  DestroyOneByIdModel,
  ReadMany,
  ReadOneById,
  ReadOneByIdModel,
  ReadOneByKeys,
  ReadOneByKeysModel,
  RestoreOneByIdModel,
} from "./database";
import { ObjectValidators, Validator } from "./validation";

export type FunctionHandlerEvent = PartialWithRequired<
  APIGatewayProxyEvent,
  "httpMethod" | "resource"
>;

export type FunctionHandlerResult =
  | PartialWithRequired<APIGatewayProxyResult, "statusCode">
  | APIGatewayProxyResult;

export type FunctionHandler = (
  event: FunctionHandlerEvent
) => Promise<FunctionHandlerResult>;

export interface FunctionHandlerFactoryArgs {
  name: string;
  eventValidators?:
    | ObjectValidators<FunctionHandlerEvent>
    | Validator<FunctionHandlerEvent>[];
}

export interface GetOneByIdFactoryArgs<
  Model extends HasId,
  TGetOneByIdModel extends GetOneByIdModel<Model> = GetOneByIdModel<Model>,
  TReadOneByIdModel extends ReadOneByIdModel<Model> = ReadOneByIdModel<Model>
> extends FunctionHandlerFactoryArgs {
  getOneModelMapper: (event: FunctionHandlerEvent) => TGetOneByIdModel;
  getOneModelValidators: ObjectValidators<TGetOneByIdModel>;
  readOneModelMapper: (model: TGetOneByIdModel) => TReadOneByIdModel;
  readOne: ReadOneById<Model, TReadOneByIdModel>;
  resultMapper?: Mapper<Model, unknown>;
}

export type GetOneByIdFactory<Model extends HasId = HasId> = (
  args: GetOneByIdFactoryArgs<Model>
) => FunctionHandler;

export interface GetOneByIdModel<Model extends HasId> {
  id: number;
  /**
   * If true gets deleted (soft) record
   */
  deleted?: boolean;
}

/**
 * @summary get one with keys handler model
 */
export type GetOneByKeysModel<Model extends any = any, Key = keyof Model> = {
  /**
   * @summary key or keys used to get the model
   */
  keys: [Key, unknown] | [Key, unknown][];
};

/**
 * @summary future usage
 */
export interface GetOneByKeysFactoryArgs<
  TGetOneByKeysModel extends GetOneByKeysModel
> extends FunctionHandlerFactoryArgs {
  getOneModelMapper: (event: FunctionHandlerEvent) => TGetOneByKeysModel;
  readOneModelMapper: (
    model: TGetOneByKeysModel
  ) => ReadOneByKeysModel<
    TGetOneByKeysModel extends GetOneByKeysModel<infer Model> ? Model : any
  >;
  readOne: ReadOneByKeys<
    TGetOneByKeysModel extends GetOneByKeysModel<infer Model> ? Model : any
  >;
}

export type GetOneByKeysFactory<Model extends HasId = HasId> = (
  args: GetOneByKeysFactoryArgs<GetOneByKeysModel<Model>>
) => FunctionHandler;

export interface GetManyHandlerFactoryArgs<GetManyModel, OneModel>
  extends FunctionHandlerFactoryArgs {
  name: string;
  getManyModelMapper: (event: FunctionHandlerEvent) => GetManyModel;
  getManyModelValidation: ObjectValidators<GetManyModel>;
  readMany: ReadMany<GetManyModel>;
}

export type DeleteOneByIdModel = {
  id: number;
};

export interface DeleteOneByIdFactoryArgs<
  TDeleteOneByIdModel extends DeleteOneByIdModel,
  TDestroyOneByIdModel extends DestroyOneByIdModel
> extends FunctionHandlerFactoryArgs {
  deleteOneByIdModelMapper: (
    event: FunctionHandlerEvent
  ) => TDeleteOneByIdModel;
  deleteOneByIdModelValidators?:
    | ObjectValidators<TDeleteOneByIdModel>
    | Validator<TDeleteOneByIdModel>[];
  destroyOneByIdModelMapper: (
    model: TDeleteOneByIdModel
  ) => TDestroyOneByIdModel;
  destroyOneById: DestroyOneById<TDestroyOneByIdModel>;
  destroyOneByIdResultMapper?: (count: number) => unknown;
}

export type UnDeleteOneByIdModel = {
  id: number;
};

/**
 * @TODO Rename to PostRestoreOneByIdFactoryArgs
 */
export interface UnDeleteOneByIdFactoryArgs<
  TUnDeleteOneByIdModel extends UnDeleteOneByIdModel,
  TRestoreOneByIdModel extends RestoreOneByIdModel
> extends FunctionHandlerFactoryArgs {
  unDeleteOneByIdModelMapper: (
    event: FunctionHandlerEvent
  ) => TUnDeleteOneByIdModel;
  unDeleteOneByIdModelValidators?:
    | ObjectValidators<TUnDeleteOneByIdModel>
    | Validator<TUnDeleteOneByIdModel>[];
  unDestroyOneByIdModelMapper: (
    model: TUnDeleteOneByIdModel
  ) => TRestoreOneByIdModel;
  unDestroyOneById: DestroyOneById<TRestoreOneByIdModel>;
  unDestroyOneByIdResultMapper?: (count: number) => unknown;
}
