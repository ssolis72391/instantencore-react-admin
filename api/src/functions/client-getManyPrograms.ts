"use strict";
import { ProgramStatus } from "api-shared";
import { APIGatewayProxyEvent } from "aws-lambda";
import { Page, Program } from "../database/models";

class ProgramsTabClient {
  title: string;
  programSummaries: ProgramSummaryClient[];
}

class ProgramSummaryClient {
  id: number;
  header: ProgramSummaryHeaderClient;
}

class ProgramSummaryHeaderClient {
  pretitle?: string;
  title?: string;
  subtitle?: string;
  image?: string;
  contentPosition?: string;
  imageSize?: string;
}

class ProgramsFiltered {
  current: Program[];
  upcoming: Program[];
  past: Program[];
  pages: Page[];
}

export const handler = async (event: APIGatewayProxyEvent) => {
  const cid = +event.queryStringParameters.cid;
  const programsFiltered = await getProgramsFiltered(cid);
  const programTabs: ProgramsTabClient[] = [];
  programTabs.push(
    createProgramTab(
      programsFiltered.current,
      "Current",
      programsFiltered.pages
    )
  );
  programTabs.push(
    createProgramTab(
      programsFiltered.upcoming,
      "Upcoming",
      programsFiltered.pages
    )
  );
  programTabs.push(
    createProgramTab(programsFiltered.past, "Past", programsFiltered.pages)
  );
  const response = {
    statusCode: 200,
    body: JSON.stringify(programTabs),
    headers: {
      "Access-Control-Allow-Headers": "Content-Type",
      "Access-Control-Allow-Origin": "*",
      /**
       * @question post for client???
       */
      "Access-Control-Allow-Methods": "OPTIONS,GET,POST",
    },
  };
  return response;
};

/**
 * Get programs (and associated pages) from the database. Group them into current, upcoming, and past.
 */
const getProgramsFiltered = async (cid: number) => {
  // Get all programs where status == published
  const programs = await Program.findAll({
    where: {
      status: ProgramStatus.published,
      clientId: cid,
    },
  });
  // Get associated pages
  const ids = programs.map((p) => p.id);
  const pages = await Page.findAll({
    where: {
      id: ids,
    },
  });
  const utcNow = new Date().getTime();
  const dayInMs = 86400000;
  // filter for current (endDateUtc.addDays(1) > utcNow) && (startDateUtc < utcNow)
  const current = programs.filter((p) => {
    const utcEndDate = new Date(p.utcEndDate).getTime() + dayInMs;
    const utcStartDate = new Date(p.utcStartDate).getTime();
    return utcEndDate > utcNow && utcStartDate < utcNow;
  });
  // filter for upcoming (startDateUtc > utcNow)
  const upcoming = programs.filter((p) => {
    const utcStartDate = new Date(p.utcStartDate).getTime();
    return utcStartDate > utcNow;
  });
  // filter for past (endDateUtc.addDays(1) < utcNow)
  const past = programs.filter((p) => {
    const utcEndDate = new Date(p.utcEndDate).getTime() + dayInMs;
    return utcEndDate < utcNow;
  });
  // grab all pages with ids in the list of ids we selected
  const programsFiltered: ProgramsFiltered = {
    current: current,
    upcoming: upcoming,
    past: past,
    pages: pages,
  };
  return programsFiltered;
};

/**
 * Transform the database data to a ProgramTabClient that will be returned.
 * @param programs Array of grouped programs (current, upcoming, or past)
 * @param title Title of the tab.
 * @param pages All associated pages so that we can retrieve header info from database
 */
const createProgramTab = (
  programs: Program[],
  title: string,
  pages: Page[]
) => {
  const programSummaries = programs.map((program) => {
    const page = pages.find((m) => m.id === program.id);
    const programSummary: ProgramSummaryClient = {
      id: program.id,
      header: {
        contentPosition: page.headerTextPosition,
        image: page.headerImageUrl,
        imageSize: page.headerImageSize,
        pretitle: page.headerPreTitle,
        subtitle: page.headerSubTitle,
        title: page.headerTitle,
      },
    };
    return programSummary;
  });
  const programTabClient: ProgramsTabClient = {
    title: title,
    programSummaries: programSummaries,
  };
  return programTabClient;
};
