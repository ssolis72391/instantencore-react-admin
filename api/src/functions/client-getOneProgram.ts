"use strict";
import { ComponentCardHelper } from "api-shared";
import { APIGatewayProxyEvent } from "aws-lambda";
import { Component, Page, PageComponent, Program } from "../database/models";
import { Card } from "../database/models/Card";

class PageDataWithNested {
  page: Page;
  components: Component[];
  cards: Card[];
}

class ActionClient {
  type: string;
  value: any;
}

class CardClient {
  id: number;
  image: string;
  title: string;
  description: string;
  action: ActionClient;
}

class ComponentClient {
  id: number;
  cardType: string;
  style: string;
  title: string;
  subtitle: string;
  cardImageSize: string;
  cardImagePosition: string;
  maxCards: number;
  viewAllText: string;
  cards: CardClient[];
}

class HeaderClient {
  pretitle: string;
  title: string;
  subtitle: string;
  image: string;
  contentPosition: string;
  imageSize: string;
}

class PageClient {
  id: number;
  title: string;
  pageType: string;
  header: HeaderClient;
  components: ComponentClient[];
}
class ProgramClient {
  id: number;
  title?: string;
  startDate?: string;
  endDate?: string;
  pages?: PageClient[];
}

export const handler = async (event: APIGatewayProxyEvent) => {
  const id = +event.pathParameters.id;
  // get data from DB
  const data = await getData(id);
  const rootPage = data.pages.find((page) => page.page.id === id);
  // map data to the client model
  const programClient: ProgramClient = {
    id: data.program.id,
    endDate: data.program.localEndDate,
    startDate: data.program.localStartDate,
    title: rootPage.page.internalName,
    pages: [],
  };
  data.pages.forEach((t) => {
    const page = t.page;
    const pageClient: PageClient = {
      id: page.id,
      title: page.internalName,
      pageType: page.id === rootPage.page.id ? "root" : "nested",
      header: {
        pretitle: page.headerPreTitle,
        title: page.headerTitle,
        subtitle: page.headerSubTitle,
        image: page.headerImageUrl,
        contentPosition: page.headerTextPosition,
        imageSize: page.headerImageSize,
      },
      components: [],
    };
    t.components.forEach((component) => {
      const componentClient: ComponentClient = {
        id: component.id,
        cardType: ComponentCardHelper.getClientCardType(component.type),
        style: component.style,
        title: component.title,
        subtitle: component.subTitle,
        cardImageSize: component.cardImageSize,
        cardImagePosition: component.cardImagePosition,
        maxCards: component.maxCards,
        viewAllText: component.viewAllText,
        cards: [],
      };
      t.cards.forEach((card) => {
        if (card.componentId == component.id) {
          const cardClient: CardClient = {
            id: card.id,
            image: card.imageUrl,
            title: card.title,
            description: card.description,
            action: null,
          };
          if (card.action != "none") {
            cardClient.action = {
              type: card.action, //TODO: verify these match
              value: card.action == "page" ? card.pageId : card.url,
            };
          }
          componentClient.cards.push(cardClient);
        }
      });
      pageClient.components.push(componentClient);
    });
    programClient.pages.push(pageClient);
  });

  const response = {
    statusCode: 200,
    body: JSON.stringify(programClient),
    headers: {
      "Access-Control-Allow-Headers": "Content-Type",
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Methods": "OPTIONS,GET,POST",
    },
  };
  return response;
};

const getData = async (id: number) => {
  const program = await Program.findOne({
    where: {
      id: id,
    },
  });
  const pages: PageDataWithNested[] = [];
  const page = await getPageDataWithNested(id);
  pages.push(page);
  await getSubPage(pages, page);

  const data = {
    program: program,
    pages: pages,
  };
  return data;
};

const getSubPage = async (pages: any[], page: PageDataWithNested) => {
  for (let i = 0; i < page.cards.length; i++) {
    const card = page.cards[i];
    if (card.pageId) {
      const pageExists = pages.find((page) => page.id === card.pageId) != null;
      if (!pageExists) {
        console.log("page exists: false");
        const subPage = await getPageDataWithNested(card.pageId);
        pages.push(subPage);
        console.log("subPage", subPage);
        await getSubPage(pages, subPage);
      } else {
      }
    }
  }
};

async function getPageDataWithNested(pageId: number) {
  // TODO: use proper associations with sequelize to retrieve data. This is hacky, but does retrieve the right data.

  // Get page
  const page = await Page.findOne({
    where: {
      id: pageId,
    },
  });

  // Get components for the page
  const pageComponents = await PageComponent.findAll({
    where: {
      pageId: pageId,
    },
    order: [["orderIndex", "ASC"]],
  });
  const componentIds = pageComponents.map(
    (pageComponent) => pageComponent.componentId
  );
  const components = await Component.findAll({
    where: {
      id: componentIds,
      deletedAt: null,
      visible: 1,
    },
  });
  let componentsOrdered = pageComponents.map((m) => {
    return components.find((c) => c.id == m.componentId);
  });
  // some may be null since we are filtering out if deletedAt is not null or visible is not 1.
  componentsOrdered = componentsOrdered.filter((m) => m != null);

  // Get cards for the components
  const cards = await Card.findAll({
    where: {
      componentId: componentIds,
      visible: 1,
    },
    order: [
      ["componentId", "ASC"],
      ["orderIndex", "ASC"],
    ],
  });

  const data: PageDataWithNested = {
    page: page,
    components: componentsOrdered,
    cards: cards,
  };
  return data;
}
