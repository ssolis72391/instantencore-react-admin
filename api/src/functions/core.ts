import { ImagePayload, rfc3339_dateonly_format } from "api-shared";
import { APIGatewayProxyResult } from "aws-lambda";
import {
  defined,
  getWrappedHandler,
  identifier,
  Logger,
  loggerFactory,
  ofType,
  rfc3339fullDate,
  storeImage,
  truthy,
  validate,
  ValidationErrors,
  Validator,
  within,
} from "../core";
import {
  validateAPIGatewayProxyEvent,
  validateAPIGatewayProxyEventModel,
  validateBody,
} from "../validation";
import dayjs = require("dayjs");

export {
  validateAPIGatewayProxyEvent,
  validateAPIGatewayProxyEventModel,
  validateBody,
};
export { defaultHeaders, DefaultHeaders, getWrappedHandler };

type DefaultHeaders = {
  "Access-Control-Allow-Headers"?: string;
  "Access-Control-Allow-Origin"?: string;
  "Access-Control-Allow-Methods"?: string;
};

const defaultHeaders: DefaultHeaders = {
  "Access-Control-Allow-Headers": "Content-Type",
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "OPTIONS,GET",
};

export function putOkResponse(body: unknown): APIGatewayProxyResult {
  const headers: DefaultHeaders = {
    ...defaultHeaders,
    "Access-Control-Allow-Methods": "GET,OPTIONS,PUT",
  };

  return {
    body: JSON.stringify(body),
    statusCode: 200,
    headers,
  };
}

export function postOkResponse(body: any): APIGatewayProxyResult {
  const headers: DefaultHeaders = {
    ...defaultHeaders,
    "Access-Control-Allow-Methods": "POST",
  };

  return {
    body: JSON.stringify(body),
    statusCode: 200,
    headers,
  };
}

export function notFound(): APIGatewayProxyResult {
  return {
    statusCode: 404,
    body: JSON.stringify({
      errorMessage: "A resource was not found",
      errorType: "NotFoundError",
    }),
    headers: defaultHeaders,
  };
}

/**
 * @todo rename to okResult
 * @param body payload to be sent as JSON in the body
 * @param loggerOrName (optional) logger or logger name that will handle the messages. If falsy an internal logger will be used.
 * @returns an `APIGatewayProxyResult` instance
 */
export function okResult(
  body: unknown,
  loggerOrName?: Logger | string
): APIGatewayProxyResult {
  const logger = !loggerOrName
    ? loggerFactory("okResult")
    : typeof loggerOrName === "object" &&
      typeof loggerOrName.debug === "function"
    ? loggerOrName
    : loggerFactory(loggerOrName);
  logger.debug("Returning APIGatewayProxyResult");
  return {
    statusCode: 200,
    body: JSON.stringify(body),
    //  todo: check if we don't need more headers or augment them
    headers: {
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Methods": "OPTIONS,GET",
    },
  };
}

export function deletedResponse(body: unknown): APIGatewayProxyResult {
  const headers: DefaultHeaders = {
    ...defaultHeaders,
    "Access-Control-Allow-Methods": "DELETE",
  };

  return {
    body: JSON.stringify(body),
    statusCode: 200,
    headers,
  };
}

/**
 * @deprecated
 */
export const handleImagePayload = async ({
  imageUrl: payloadImageUrl,
  isImageDataUrl,
}: ImagePayload) =>
  isImageDataUrl === true
    ? await storeImage(payloadImageUrl)
    : payloadImageUrl || null;

export const HttpMethods = {
  PUT: "PUT",
  POST: "POST",
  DELETE: "DELETE",
  GET: "GET",
  OPTIONS: "OPTIONS",
};

export type ModelHandling<Model> =
  | ((model: Model) => void | any)
  | ((model: Model) => void | any)[];

// interface HandlerFactoryArgs {
//   eventValidators:
//     | Validator<APIGatewayProxyEvent>
//     | Validator<APIGatewayProxyEvent>[];
//   name: string;
// }

interface AcceptsDalResultValidator<Model> {
  dalResultValidators?: Validator<Model> | Validator<Model>[];
  /**
   * Future usage
   */
  dalResultValidatorResultsHandlers?: (
    errors: ValidationErrors | ValidationErrors[]
  ) => void | APIGatewayProxyResult;
}

export function createdResult(body: unknown): APIGatewayProxyResult {
  const headers: DefaultHeaders = {
    ...defaultHeaders,
    "Access-Control-Allow-Methods": "POST",
  };

  return {
    body: JSON.stringify(body),
    statusCode: 201,
    headers,
  };
}

export function modelCreatedResult(id: number) {
  validate("id", "argument", id, identifier, "modelCreatedResult");
  return createdResult({ id });
}

/**
 * @todo
 * - validate image extension
 * - validate image data url is well formed
 * - validate image is not local
 */
export async function storeAndOrGetImageUrl(imageUrl: string): Promise<string> {
  if (imageUrl.startsWith("http://") || imageUrl.startsWith("https://")) {
    return imageUrl;
  } else {
    return await storeImage(imageUrl);
  }
}

//#region todo: type these!
export const defaultOneModelMapper = (event) => {
  const id = +event.pathParameters.id;
  return {
    id,
  } as const;
};

export const defaultGetOneModelValidators = { id: truthy() } as const;
export const defaultReadOneModelMapper = ({ id }) => ({ id } as const);
//#endregion

export function toUtc(value: string, offset: number) {
  validate("value", "argument", value, [truthy(), rfc3339fullDate]);
  validate("offset", "argument", offset, [
    defined(),
    ofType("number"),
    within(-11, 14),
  ]);

  const date = dayjs(value, rfc3339_dateonly_format, true);
  date.add(offset, "hours");

  return date.format(rfc3339_dateonly_format);
}
