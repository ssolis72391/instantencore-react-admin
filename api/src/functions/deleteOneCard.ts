import { equals, getWrappedHandler, truthy } from "../core";
import { Card } from "../database/models";
import { deleteOneByIdFactory } from "./factories/deleteOneByIdFactory";

export const deleteOneCard = deleteOneByIdFactory({
  name: "deleteOneCard",
  eventValidators: {
    resource: equals("/cards/{id}"),
    pathParameters: truthy(),
  },
  deleteOneByIdModelMapper: (event) => ({ id: +event.pathParameters.id }),
  destroyOneByIdModelMapper: ({ id }) => ({ id }),
  destroyOneById: async (model) => {
    const result = await Card.destroy({
      where: {
        id: model.id,
      },
    });
    return result;
  },
});

export const handler = getWrappedHandler(deleteOneCard);
