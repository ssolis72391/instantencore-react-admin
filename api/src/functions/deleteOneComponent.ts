import { equals, getWrappedHandler, truthy } from "../core";
import { deleteOneByIdFactory } from "./factories/deleteOneByIdFactory";

export const deleteOneComponent = deleteOneByIdFactory({
  name: "deleteOneComponent",
  eventValidators: {
    resource: equals("/components/{id}"),
    pathParameters: truthy(),
  },
  deleteOneByIdModelMapper: (event) => ({ id: +event.pathParameters.id }),
  destroyOneByIdModelMapper: ({ id }) => ({ id }),
  destroyOneById: (model) => Promise.resolve(1),
});

export const handler = getWrappedHandler(deleteOneComponent);
