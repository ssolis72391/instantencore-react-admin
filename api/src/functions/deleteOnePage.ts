import { equals, getWrappedHandler } from "../core";
import { updateOnePage } from "../database/updateOnePage";
import { deleteOneByIdFactory } from "./factories/deleteOneByIdFactory";

export const deleteOnePage = deleteOneByIdFactory({
  name: "deleteOnePage",
  eventValidators: {
    resource: equals("/pages/{id}"),
  },
  deleteOneByIdModelMapper: ({ pathParameters }) => ({
    id: +pathParameters.id,
  }),
  destroyOneByIdModelMapper: ({ id }) => ({ id }),
  destroyOneById: ({ id }) =>
    updateOnePage({ id, model: { status: "deleted" } }),
});

export const handler = getWrappedHandler(deleteOnePage);
