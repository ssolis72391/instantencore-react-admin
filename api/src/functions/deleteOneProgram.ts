import { equals, getWrappedHandler } from "../core";
import { detroyOneProgram } from "../database/detroyOneProgram";
import { deleteOneByIdFactory } from "./factories/deleteOneByIdFactory";

/**
 * @todo fix event validation
 */
export const deleteOneProgram = deleteOneByIdFactory({
  name: "deleteOneProgram",
  eventValidators: {
    resource: equals("/programs/{id}"),
  },
  deleteOneByIdModelMapper: (event) => ({ id: +event.pathParameters.id }),
  destroyOneByIdModelMapper: ({ id }) => ({ id }),
  destroyOneById: detroyOneProgram,
});

export const handler = getWrappedHandler(deleteOneProgram);
