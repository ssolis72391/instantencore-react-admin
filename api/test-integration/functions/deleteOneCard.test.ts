import * as AWS from "aws-sdk";
import * as AWSMock from "aws-sdk-mock";
import * as LambdaTester from "lambda-tester";
import { handler } from "../../src/functions/deleteOneCard";

AWSMock.setSDKInstance(AWS);

/**
 * TODO: Until we have a proper seed method, you will need to update `pathParameters.id` to match a value in your database.
 * You will also need to update your snapshot if you change the id. (add `-u` to command)
 */
describe("deleteOneCard", () => {
  test("deleteOneCard response", async () => {
    expect.assertions(2);
    await LambdaTester(handler)
      .event({
        httpMethod: "DELETE",
        pathParameters: { id: "20" },
        resource: "/cards/{id}",
      })
      .timeout(10)
      .expectResolve((result) => {
        // Snapshot whole response, allowing any body, to be checked next.
        expect(result).toMatchSnapshot({
          body: expect.any(String),
        });

        // Testing body
        expect(JSON.parse(result.body)).toMatchSnapshot({
          id: expect.any(Number),
        });
      });
  });
});
