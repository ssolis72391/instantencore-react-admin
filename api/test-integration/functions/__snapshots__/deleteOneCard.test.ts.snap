// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deleteOneCard deleteOneCard response 1`] = `
Object {
  "body": Any<String>,
  "headers": Object {
    "Access-Control-Allow-Headers": "Content-Type",
    "Access-Control-Allow-Methods": "DELETE",
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`deleteOneCard deleteOneCard response 2`] = `
Object {
  "id": Any<Number>,
}
`;
